### ISSUE TITLE
mrouted 4.0 reports "Too many files on filesystem"

### ISSUE THREAD

**Issue by F-i-f** [troglobit/mrouted#40]
I have mrouted running on a Linux router sitting between multiple networks, only two of which have IFF_MULTICAST.
mrouted is used so that that multicast can be properly forwarded between the two multicast-enabled networks.
mrouted 4.0 runs without a configuration file (or rather, the empty, all-commented out /etc/mrouted.conf).
Every once in a while, mrouted will log (public IP address redacted as `a.b.c.d`:
```
mrouted[276475]: warning - Failed MRT_ADD_MFC(a.b.c.d, 239.255.255.250) from vif 32 to vif(s) 0, 0: Too many open files in system
mrouted[276475]: warning - Failed MRT_DEL_MFC(a.b.c.d 239.255.255.250): No such file or directory
mrouted[276475]: warning - age_table_entry() trying to delete no-route (a.b.c.d 239.255.255.250): No such file or directory
```
Looking at mrouted and the Linux kernel's code it looks like I'm running out of multicast virtual interfaces (vifs).
MAXVIFS is defined as 32 in my stock kernel.
I have 18 interfaces configures on the router (2 of which have IFF_MULTICAST).

**Comment by troglobit** [comment]
At first glance, this looks like a duplicate of #36. But since only two interfaces have `IFF_MULTICAST` set, this could be something else, because vif 32 is usually used for error codes/detection. Also, group 239.255.255.250 is a link-local group which usually has TTL 1.

Could you include the output from `/proc/net/ip_mr_vif`?

**Comment by F-i-f** [comment]
Sure:
```
Interface      BytesIn  PktsIn  BytesOut PktsOut Flags Local    Remote
 0 br0        67006834  130407    187900     657 00000 010111AC 00000000
 1 enp8s0.10   2915176   16467    709550    1380 00000 010211AC 00000000
```

**Comment by troglobit** [comment]
As I suspected then, we only have two vifs. So your error is something else. Looks weird.

Maybe you can run mrouted with debug enabled to get more context for the error messages?

**Comment by F-i-f** [comment]
Sounds good.
I'm running mrouted with `/usr/sbin/mrouted -ns -d all`.
And I found a way to reproduce the problem:
I'm just subscribing to 231.1.1.1 on a host attache to the br0 interface, and I get:
```
mrouted[644817]: debug level 0xffffffff (packet, pkt, pruning, prunes, routing, routes, route-detail, rtdetail, neighbors, peers, kernel, cache, timer, vif, interface, groups, membership, mtrace, traceroute, igmp, icmp, rsrr, all, 3)
mrouted[644817]: mrouted version 4.0
mrouted[644817]: warning - Failed MRT_ADD_MFC(107.194.242.74, 231.1.1.1) from vif 32 to vif(s) 0, 0: Too many open files in system
mrouted[644817]: warning - Kernel entry already exists for (107.194.242.74 231.1.1.1)
mrouted[644817]: warning - Failed MRT_ADD_MFC(107.194.242.74, 231.1.1.1) from vif 32 to vif(s) 0, 0: Too many open files in system
```

I was able to run mrouted with `strace -s 512 -o mrouted.strace.txt /usr/sbin/mrouted -n -d all` and I've attached the [resulting log file](https://github.com/troglobit/mrouted/files/5277353/mrouted.strace.txt).  strace was complaining about not being able to fully decode netlink messages (I think):


```
strace: decode_nlattr: [xlat 0x561a4ae856a0, dflt "AF_???", decoders 0x7ffcdf756330] size is zero (going to pass nla_type as decoder argument), but opaque data (0x7ffcdf7563f0) is not - will be ignored
```
I couldn't see any extra logging from the debug flag, and nothing obvious comes out of the strace output...

**Comment by troglobit** [comment]
Cool thanks, I'll have a look at it! You can also add `-l debug` to get more logs

**Comment by F-i-f** [comment]
> Cool thanks, I'll have a look at it! You can also add `-l debug` to get more logs

Sorry, I missed `-l debug`.

Are you able to reproduce the problem yourself, or would you need another log file with `-l debug`?

**Comment by troglobit** [comment]
No worries. It would be great if you could run it, not sure if I'll be able to do so reliably myself, and if I do it's always good to have a reference.

**Comment by F-i-f** [comment]
There you go:
[mrouted.debug.log](https://github.com/troglobit/mrouted/files/5279199/mrouted.debug.log)

**Comment by troglobit** [comment]
Thank you so much!

**Comment by troglobit** [comment]
Took a while, but I think I may know what's going on here a bit better now.  The error message you get; "Too many open files in system" comes from the kernel when mrouted tries to create a kernel route from vif 32, which as I mentioned previously is reserved for error handling.  It comes from here:

https://github.com/troglobit/mrouted/blob/61767bad5092ccb03798a7dff519c470fcbe72be/src/kern.c#L243

Linux, unlike BSD where mrouted was written for originally, barfs if the parent interface is `NO_VIF` (32).  So from this we can conclude that at the time of the log message:

https://github.com/troglobit/mrouted/blob/61767bad5092ccb03798a7dff519c470fcbe72be/src/kern.c#L257

mrouted doesn't have the reverse path for the source address.  We end up in `k_add_rg()` and this error message due to the kernel calling `add_table_entry()` when a new multicast stream without a route appears. The kernel asks mrouted "Hey, here's incoming data from source address S, to group G.  What do you want to do with it?"  The `add_table_entry()` function calls `determine_route()` to figure out the inbound interface, but that function returns `NULL` and here we are.

When you get this, you can check with the command:

    mroutectl -d show mfc

You should get an entry listed with `!!` in the flags field `<>`, to denote an incoming stream without a route.  This is possibly more intuitively listed in `mroutectl show compat` (the old debug/status output).

I have no idea atm why you get this.  I've tried reproducing your setup, but no luck so far. :disappointed:

**Comment by F-i-f** [comment]
Well my setup is a little more complicated as some hosts are multihomed with more than 2 IPs on the same interface.
And that was messing things up and your explanations made me figure out what's wrong.

As expected, when the problem manifested itself, `mroutectl -d show mfc` showed:
```
Multicast Forwarding Cache Table
11.11.11.130/32 231.1.1.1       -               !!   0:00:11  0:05:55
172.17.2.0/26   239.255.255.250 enp8s0.10            1:47:00  0:01:30  br0
172.17.1/24     239.255.255.250 br0                 11:26:26  0:06:36

Source          Group           Inbound               Uptime  Packets     Bytes 
172.17.2.5      239.255.255.250 enp8s0.10            1:47:00      216     43200
172.17.2.11     239.255.255.250 enp8s0.10            0:11:10       94     11656
172.17.1.1      239.255.255.250 br0                 11:26:26    32952  16942820
```

Let's step back and give you more details.  On the router running mrouted, I have:
```
172.17.1.1/24, IFF_MULTICAST on br0
11.11.11.129/25 as an alias on br0 (not the real IP, but a routable one)
172.17.2.1/26, IFF_MULTICAST enp8s0.10
```

On the machine that causes the problem, I have:
```
172.17.1.2/24, IFF_MULTICAST on br0
11.11.11.130/25 as an alias on br0
```

And the routing table looked like:
```
default via 11.11.11.129 dev br0 proto static metric 425 
11.11.11.128/25 dev br0 proto kernel scope link src 11.11.11.130 metric 425 
11.11.11.129 dev br0 scope link src 11.11.11.130 metric 415 
172.17.0.0/16 via 172.17.1.1 dev br0 proto static metric 425 
172.17.1.0/24 dev br0 proto kernel scope link src 172.17.1.2 metric 425 
```
Because of the routing table, talking to the internal network on 172.17.0.0/16 will lead the machine to use its private IP address.
But for anything else, the public IP 11.11.11.130/25 will be used.  Including for multicast traffic.
Just adding a route for 224.0.0.0/4 makes the problem go away:
`224.0.0.0/4 dev br0 scope link metric 425`

So while the problem was definitely on my side, maybe mrouted could emit a better error message (something akin to _multicast packet received on non-multicast interface_).

**Comment by troglobit** [comment]
Interesting, thank you for the follow-up!

We should definitely emit a better error message in these cases, that's the bare minimum!  

The root cause here is that mrouted (and DVMRP) is really stupid.  It builds its own routing table for the reverse path based on the interfaces and their address+netmask, doesn't even consult the regular routing table.  It's old and it shows ... but since we still like it, I'm thinking about how to fix cases like yours.  Even though they're perhaps not common, your setup is definitely not unique.  We provide similar setups with our products at $DAYJOB.

Question, is it another IP address, set with `ip addr`, or an alias interface to br0?  I'd like to see what `getifaddrs()` returns (on Linux and *BSD) to see if we can complement the mrouted DVRMP routing table.

**Comment by troglobit** [comment]
I've added a detection and more appropriate log message in 1e1cbf8.  Keeping this issue open to see if I can figure out something else, e.g., multiple networks per interface support.

**Comment by troglobit** [comment]
Found it!  Turns out we already had support for IP aliases in mrouted.  They're called 'altnet' and are configured on each phyint.  In 48a7a11 I added automatic detection and install of altnets for secondary IP addresses.

Thanks for hanging in there and the excellent bug report(s)! :smiley::+1:

**Comment by F-i-f** [comment]
> Question, is it another IP address, set with `ip addr`, or an alias interface to br0? I'd like to see what `getifaddrs()` returns (on Linux and *BSD) to see if we can complement the mrouted DVRMP routing table.

That's an extra address (ip addr add), not an alias interface.  Sorry about mentioning aliasing earlier on and getting everyone confused :-(

Here's how to reproduce this consistently.
Tried it with some VMs, and it works.
On the router system:
```
dev1=enp1s0
dev2=enp2s0
ip addr add 192.168.0.1/24 dev $dev1
ip addr add 10.0.0.1/24 dev $dev1
ip link set $dev1 up
ip addr add 192.168.1.1/24 dev $dev2
ip link set $dev2 up
ip route add 224.0.0.0/4 dev $dev1
```

On the satellite system:
```
dev=enp1s0
ip addr add 192.168.0.2/24 dev $dev
ip addr add 10.0.0.2/24 dev $dev
ip link set $dev up
ip route add default via 10.0.0.1 src 10.0.0.2
```

The router's enp1s0 must be connected to the satellite's enp1s0.

Then on router, run `mrouted -n`.
And on the satellite, subscribe to any multicast group.

With this process, I can reliably reproduce the original error message with mrouted 4.0, which comes up within a second of the group being subscribed on the satellite system:
```
warning - Failed MRT_ADD_MFC(192.168.0.2, 231.1.1.1) from vif 32 to vif(s) 0, 0: Too many open files in system
```
I've also tried with the latest master snapshot (8a93ce), and I was expecting to see the new error message, but instead I get:
```
wwarning - Kernel entry already exists for (192.168.0.2 231.1.1.1)
```
a few seconds after the group has been subscribed, and repeats at a regular interval.

Of course, if the I add an on-link route for `224.0.0.0/4` on the satellite, everything works fine.

**Comment by troglobit** [comment]
No worries, I think under the hood Linux treats aliases and extra addresses the same these days :smiley: 

The new error message was always there, just another special case to the same buggy handling.  The "auto altnet" change I submitted (48a7a11), a few commits after the master snapshot you tested, handles this better.  

I've tested and re-tested several times now, using a second IP address on one of my router's interfaces.  Before the change I can reproduce your bug, after the change it works perfectly :sunglasses:

**Comment by F-i-f** [comment]
Confirmed that the latest master (221966) works in the configuration quoted above.
Thanks.

The same latest master has a problem with regards to autotools expansion:
```shell
$ strings src/mrouted | grep '^@'
@runstatedir@/mrouted.sock
@runstatedir@
```

And running that `mrouted` binary yields:
```
warning - Failed binding IPC socket, @runstatedir@/mrouted.sock: No such file or directory
```

While I'm here, I'll be testing multicast routing with IPv6, in a similar set up (only trying to route between two local subnets, no tunnels or upstream multicast routing).
I don't think `mrouted` supports IPv6, but I may be wrong.  Maybe that's worth mentioning somewhere on the project page or manual.
Since you seem to be maintaining most multicast routing software these days (thanks!), would there be any daemon you can recommend?  I'll discount `smcroute` as I'd prefer dynamic routing.

**Comment by troglobit** [comment]
Thank you for taking the time to test and give feedback, it's most appreciated!

Ah, the runstatedir problem.  Ran into this in SMCRoute as well, appears GNU hasn't officially released that feature yet, but most major distros (RedHat/Debian) have backported it to their autoconf packages.  I'll get right on with fixing that, it's a simple workaround, thanks!

You're right, mrouted is IPv4 only.  I'm blind to these sort of details since I've work with it for so long.  I'll update the README and possibly the man page as well. 

Wrt multicast routing and IPv6, there's https://github.com/troglobit/pim6sd ... dunno about the state of PIM in FRR, but I know they now have SM and SSM support for IPv4 at least.  (They started out with qpimd, from Quagga, which was SSM-only).  The pim6sd project is in desperate need of attention, debugging/testing ... much of the code is the same as the old pimd, but since it's based on an older version there may very well be odd system bugs that are fixed in pimd but not in pim6sd. For instance, I've already started fixing up the netlink backend, but I guess it's far from ready.  Post in the issues section, maybe that'll wake up @T-X as well ;-)
